pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,387 @@
+        (c_void_p, 'conduit_node_set_path_int64', c_void_p, c_char_p,
+         c_longlong),
+        (c_void_p, 'conduit_node_set_path_int64_ptr', c_void_p, c_char_p,
+         c_void_p, c_longlong),
+        (c_void_p, 'conduit_node_set_path_node', c_void_p, c_char_p, c_void_p),
+        (c_void_p, 'conduit_node_print', c_void_p),
+    ]
+
+    def _errcheck(self, status, fn, args):
+        if not status:
+            raise ConduitError
+
+
+class ConduitNode:
+    def __init__(self, lib, ptr=None, child=False):
+        self.lib = lib
+        self.child = child
+        self._as_parameter_ = ptr or self.lib.conduit_node_create(None)
+
+    def __del__(self):
+        if not self.child:
+            self.lib.conduit_node_destroy(self)
+
+    def __setitem__(self, key, value):
+        key = key.encode()
+        match value:
+            case str():
+                self.lib.conduit_node_set_path_char8_str(self, key,
+                                                         value.encode())
+            case ConduitNode():
+                self.lib.conduit_node_set_path_node(self, key, value)
+            case int():
+                self.lib.conduit_node_set_path_int64(self, key, value)
+            case float():
+                self.lib.conduit_node_set_path_float64(self, key, value)
+            case (np.ndarray() | np.generic()):
+                value = np.ascontiguousarray(value)
+                fn = getattr(self.lib,
+                             f'conduit_node_set_path_{value.dtype}_ptr')
+                fn(self, key, value.ctypes.data, value.size)
+            case list():
+                value = np.array(value, dtype=float)
+                self.lib.conduit_node_set_path_float64_ptr(self, key,
+                                                           value.ctypes.data,
+                                                           value.size)
+            case _:
+                ValueError('ConduitNode: __setitem__ type not supported')
+
+    def __str__(self):
+        self.lib.conduit_node_print(self)
+        return '\n'
+
+    def append(self):
+        ptr = self.lib.conduit_node_append(self)
+        return ConduitNode(self.lib, ptr, child=True)
+
+
+class AscentWrappers(LibWrapper):
+    _libname = 'ascent_mpi'
+
+    # Functions
+    _functions = [
+        (None, 'ascent_close', c_void_p),
+        (c_void_p, 'ascent_create', c_void_p),
+        (None, 'ascent_execute', c_void_p, c_void_p),
+        (None, 'ascent_open', c_void_p, c_void_p),
+        (None, 'ascent_publish', c_void_p, c_void_p),
+    ]
+
+
+class AscentPlugin(RegionMixin, BasePlugin):
+    name = 'ascent'
+    systems = ['*']
+    formulations = ['dual', 'std']
+
+    # Element name mapping for conduit
+    bp_emap = {'hex': 'hex', 'pri': 'wedge', 'pyr': 'pyramid', 'quad': 'quad',
+               'tet': 'tet', 'tri': 'tri'}
+
+    # Ascent filters that implicitly require velocity expression
+    v_filter = ['gradient', 'qcriterion', 'vorticity']
```

Comment:
We probably want to extend this. (Think about MHD where there are two vector fields and taking the gradients of B is not a crazy idea.)

Reply:
Use a set here.